Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs

```diff
@@ -0,0 +1,181 @@
+
+    [TestMethod]
+    public async Task LanguageServerShutsDownIfParentDies() {
+      var process = await StartLanguageServerRunnerProcess();
+
+      var languageServerProcessId = await process.StandardOutput.ReadLineAsync();
+
+      var initializeMessage = GetLspInitializeMessage(process.Id);
+      await process.StandardInput.WriteAsync(initializeMessage);
+
+      var initializedResponseFirstLine = await process.StandardOutput.ReadLineAsync();
+      Assert.IsFalse(string.IsNullOrEmpty(initializedResponseFirstLine));
+
+      var error = await process.StandardError.ReadToEndAsync();
+      var didExit = process.WaitForExit(-1);
+      Assert.IsTrue(didExit);
+      Assert.IsNotNull(languageServerProcessId, error);
+      Assert.IsTrue(string.IsNullOrEmpty(error), error);
+
+      // Wait for the language server to kill itself by waiting until it closes the output stream.
+      await process.StandardOutput.ReadToEndAsync();
+      Thread.Sleep(200); // Give the process some time to die
+      try {
+        var languageServer = Process.GetProcessById(int.Parse(languageServerProcessId));
+        languageServer.Kill();
+        Assert.Fail("Language server should have killed itself if the parent is gone.");
+      } catch (ArgumentException) {
+        // Language server process is not running, pass the test.
+      }
+    }
+
+    private static async Task<Process> StartLanguageServerRunnerProcess() {
+      var languageServerBinary = Path.Join(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), "DafnyLanguageServer");
+      var languageServerRunnerPath = await CreateDotNetDllThatStartsGivenFilepath(languageServerBinary.Replace("\\", "\\\\"));
+
+      var processInfo = new ProcessStartInfo("dotnet", languageServerRunnerPath) {
+        RedirectStandardOutput = true,
+        RedirectStandardError = true,
+        RedirectStandardInput = true,
+        UseShellExecute = false
+      };
+      return Process.Start(processInfo)!;
+    }
+
+    class AlwaysOutputFilter : IOutputFilter {
+      public bool ShouldOutput(object value) => true;
+    }
+
+    private static string GetLspInitializeMessage(int? processId) {
+      var buffer = new MemoryStream();
+      OutputHandler outputHandler = new OutputHandler(PipeWriter.Create(buffer), new JsonRpcSerializer(),
+        new List<IOutputFilter> { new AlwaysOutputFilter() },
+        TaskPoolScheduler.Default, null);
+      outputHandler.Send(new OutgoingRequest {
+        Id = 1,
+        Method = "initialize",
+        Params = new InitializeParams {
+          ProcessId = processId,
+          ClientInfo = new ClientInfo(),
+          Capabilities = new ClientCapabilities(),
+        }
+      });
+#pragma warning disable VSTHRD002
+      outputHandler.StopAsync().Wait();
+#pragma warning restore VSTHRD002
+      return Encoding.ASCII.GetString(buffer.ToArray());
+    }
+
+    private static async Task<string> CreateDotNetDllThatStartsGivenFilepath(string filePathToStart) {
```

Comment:
Since it is not explicit in the method name - and we cannot enlarge the method name - please add a comment to this method such as the following:

    // This method starts a process (like an IDE) that starts the Dafny Language Server but just crashs after reading one line.
    // It enables to test if the DafnyLanguageServer detects that its parent process died and also terminates